### PR TITLE
Add device category for daemon

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Adds a descriptive category to each device type. WIP since this should land after @DanTup 's fix to version - and then after I update the version again.

### CODE REVIEW

File: packages/flutter_tools/lib/src/device.dart [hunk 1]
```diff
@@ -28,6 +28,23 @@ import 'windows/windows_device.dart';
...
 
+/// A broad category the describes the kind of workflow the device supports.
+class Category {
+  const Category._(this.value);
+
+  static const Category web = Category._('web');
+  static const Category fuchsia = Category._('fuchsia');
+  static const Category desktop = Category._('desktop');
+  static const Category mobile = Category._('mobile');
+  static const Category simulator = Category._('simulator');
```
Comment (DanTup) [inline packages/flutter_tools/lib/src/device.dart]: Should we call this "emulator" or "simulator"? We stick the iOS simulators in the "flutter emulators" command and generally refer to them as emulators (except for their names). I don't know which type is more common though!

**Review by DanTup — COMMENTED** [review]
That was quick! :)

We should also include a note in the daemon doc + rev. the version. ~~I already have an open change modifying the changelog (https://github.com/flutter/flutter/pull/33980) but since it keeps failing bots with flakes, feel free to just include that in your PR and I'll abandon that.~~

**Comment by DanTup** [comment]
Oops, I somehow missed that you'd actually mentioned the version change in your message! Anyway, my change (https://github.com/flutter/flutter/pull/33980) is merged now so you can rebase and update.

**Comment by DanTup** [comment]
We also need to add the `platformType` field (to actually filter the devices against what's returned from `getSupportedPlatforms`). Not sure if you want to add it here, or we can do in another PR (I dont' mind doing it, but it might conflict with this, so best done afterwards if not in this one).

File: packages/flutter_tools/lib/src/device.dart [hunk 2]
```diff
@@ -28,6 +28,23 @@ import 'windows/windows_device.dart';
 
 DeviceManager get deviceManager => context.get<DeviceManager>();
 
+/// A broad category the describes the kind of workflow the device supports.
+class Category {
+  const Category._(this.value);
+
+  static const Category web = Category._('web');
+  static const Category fuchsia = Category._('fuchsia');
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/device.dart]: As per an email thread, I think we should eliminate this category and the `simulator` below, and for now just have `web`, `mobile`, and `desktop`.

File: packages/flutter_tools/lib/src/commands/daemon.dart [hunk 3]
```diff
@@ -763,12 +763,15 @@ dynamic _toEncodable(dynamic object) {
...
+  final bool isLocalEmulator = await device.isLocalEmulator;
   return <String, dynamic>{
     'id': device.id,
     'name': device.name,
     'platform': getNameForTargetPlatform(await device.targetPlatform),
-    'emulator': await device.isLocalEmulator,
+    'emulator': isLocalEmulator,
     'category': device.category.toString(),
+    'platformType': device.platformType.toString(),
+    'ephemeral': device.ephemeral ?? isLocalEmulator,
```
Comment (DanTup) [inline packages/flutter_tools/lib/src/commands/daemon.dart]: Does this result in Android mobiles (non-emulators) being non-ephemeral? That would mean we wouldn't auto-select them when you plug one in?

File: packages/flutter_tools/lib/src/ios/devices.dart [hunk 4]
```diff
@@ -114,7 +114,12 @@ class IOSDevices extends PollingDeviceDiscovery {
 class IOSDevice extends Device {
   IOSDevice(String id, { this.name, String sdkVersion })
       : _sdkVersion = sdkVersion,
-        super(id, category: Category.mobile) {
+        super(
+          id,
+          category: Category.mobile,
+          platformType: PlatformType.ios,
+          ephemeral: false,
```
Comment (DanTup) [inline packages/flutter_tools/lib/src/ios/devices.dart]: (same question as above - seems like this should be `true`?)

File: packages/flutter_tools/lib/src/web/web_device.dart [hunk 5]
```diff
@@ -28,7 +28,12 @@ class WebApplicationPackage extends ApplicationPackage {
 }
 
 class WebDevice extends Device {
-  WebDevice() : super('web', category: Category.web);
+  WebDevice() : super(
+      'web',
+      category: Category.web,
+      platformType: PlatformType.web,
+      ephemeral: true,
```
Comment (DanTup) [inline packages/flutter_tools/lib/src/web/web_device.dart]: I'm not familiar with how this one works, but I would've imagined it'd behave like desktop (eg. always be available in the device list, not appear/disappear? if so, shouldn't this be false?)

**Comment by jonahwilliams** [comment]
Updated to add category, platformType, and ephemeral flag

**Comment by jonahwilliams** [comment]
I might have misunderstood the intention behind ephemeral - is that idea this is a device that is created on demand, or was is that it is required to be connected?

**Comment by DanTup** [comment]
My understanding was that this was to distinguish devices that the user has to explicitly connect (or start), so that when we see events for them from the device daemon, we can assume they're a good candidate to auto-switch to. For desktop devices we get an event saying they were added at startup, so right now depending on the order of macOS and iPhone at startup, you may end up on macOS (in fact, it always seems to happen that way for me).

That said, there are other ways we can handle that (for ex. getting a list of devices at startup, so we know if they were really "added" or just "already there", or tweaking the order the daemon would advertise devices at startup that are already available), so if there was another use for this flag that doesn't match this, we could explore those.

**Comment by jonahwilliams** [comment]
Understood, in that case only the simulators/emulators would be ephemeral, could we instead use the existing mobile category?

File: packages/flutter_tools/doc/daemon.md [hunk 6]
```diff
@@ -250,6 +250,7 @@ See the [source](https://github.com/flutter/flutter/blob/master/packages/flutter
 
 ## Changelog
 
+- 0.5.1: Added `platformType`, `ephemeral`, and `category` field to device.
```
Comment (devoncarew) [inline packages/flutter_tools/doc/daemon.md]: We should also update the docs for the `device.getDevices` call and the `device.added` and `device.removed` events.

**Comment by DanTup** [comment]
> Understood, in that case only the simulators/emulators would be ephemeral

+ real mobiles?

> could we instead use the existing mobile category?

Potentially, but if in the future you might be able to run a "desktop emulator" I would think that would be ephemeral too? Or if you could connect to a "remote desktop" machine I'd probably consider that ephemeral (maybe it's easier to consider the opposite - those that are "always available" for the given platform - generally desktop for the local machine + web?).

File: packages/flutter_tools/lib/src/commands/daemon.dart [hunk 7]
```diff
@@ -763,11 +763,15 @@ dynamic _toEncodable(dynamic object) {
...
 
 Future<Map<String, dynamic>> _deviceToMap(Device device) async {
+  final bool isLocalEmulator = await device.isLocalEmulator;
   return <String, dynamic>{
     'id': device.id,
     'name': device.name,
     'platform': getNameForTargetPlatform(await device.targetPlatform),
-    'emulator': await device.isLocalEmulator,
+    'emulator': isLocalEmulator,
+    'category': device.category.toString(),
```
Comment (DanTup) [inline packages/flutter_tools/lib/src/commands/daemon.dart]: (haven't tested it, but this might need a `?` to allow the `null` category)

**Comment by jonahwilliams** [comment]
Update daemon docs

**Review by devoncarew — APPROVED** [review]
lgtm for my part

**Comment by DanTup** [comment]
My only outstanding comment is the one about Android devices being `ephemeral: null` and then using `isLocalEmulator`. I think it should be true (and no emulator check) so that non-emulator Android mobiles are considered ephemeral?

**Comment by jonahwilliams** [comment]
I've done a pass on the ephemeral flags and matched them to the correct values

**Review by DanTup — APPROVED** [review]
LGTM, thanks!

**Comment by DanTup** [comment]
I just noticed that in `getSupportedPlatforms` the things we call `platforms` are called `platformTypes` here.

If we want to fix this, it might be best done before we ship anything (I have no strong preference, but I thought it worth mentioning).

**Pull request merged**